lib/render/ParticleRenderer.cpp

```diff
@@ -52,46 +55,353 @@ using namespace VAPoR;
+}
+
+bool ParticleRenderer::_particleCacheIsDirty() const {
+    auto p = GetActiveParams();
+
+    if (_cacheParams.ts        != p->GetCurrentTimestep()    ) return true;
+    if (_cacheParams.rLevel    != p->GetRefinementLevel()    ) return true;
+    if (_cacheParams.cLevel    != p->GetCompressionLevel()   ) return true;
+
+    VAPoR::CoordType min, max;
+    p->GetBox()->GetExtents(min, max);
+    if (_cacheParams.boxMin != min) return true;
+    if (_cacheParams.boxMax != max) return true;
+
+    if (_cacheParams.stride    != p->GetValueLong( ParticleParams::StrideTag, 1)) return true;
+    if (_cacheParams.varName   != p->GetVariableName()       ) return true;
+    if (_cacheParams.fieldVars != p->GetFieldVariableNames() ) return true;
+    if (_cacheParams.direction != (bool)p->GetValueLong( ParticleParams::ShowDirectionTag, false)) return true;
+    if (_cacheParams.directionScale != p->GetValueDouble( ParticleParams::DirectionScaleTag, false)) return true;
+
+    return false;
+}
+
+void ParticleRenderer::_resetParticleCache() {
+    auto p = GetActiveParams();
+    _cacheParams.ts      = p->GetCurrentTimestep();
+    _cacheParams.rLevel  = p->GetRefinementLevel();
+    _cacheParams.cLevel  = p->GetCompressionLevel();
+
+    VAPoR::CoordType min, max;
+    p->GetBox()->GetExtents(min, max);
+    _cacheParams.boxMin = min;
+    _cacheParams.boxMax = max;
+   
+    _cacheParams.direction = (bool)p->GetValueLong(ParticleParams::ShowDirectionTag, false); 
+    _cacheParams.directionScale = p->GetValueDouble(ParticleParams::DirectionScaleTag, 1.);
+    _cacheParams.stride = p->GetValueLong(ParticleParams::StrideTag, 1);
+    _cacheParams.varName = p->GetVariableName();
+    _cacheParams.fieldVars = p->GetFieldVariableNames();
+}
+
+bool ParticleRenderer::_colormapCacheIsDirty() const {
+    auto p = GetActiveParams();
+    std::vector<float> tf_lut;
+    MapperFunction *tf = p->GetMapperFunc(_cacheParams.varName);
+    tf->makeLut(tf_lut);
+    if (_cacheParams.tf_lut    != tf_lut)                  return true;
+    if (_cacheParams.tf_minMax != tf->getMinMaxMapValue()) return true;
+    return false;
+}
+
+void ParticleRenderer::_resetColormapCache() {
+    auto p = GetActiveParams();
+    MapperFunction *tf = p->GetMapperFunc(_cacheParams.varName);
+    std::vector<float> tf_lut;
+    tf->makeLut(tf_lut);
+    std::vector<double> minMax = tf->getMinMaxMapValue();
+    _cacheParams.tf_lut    = tf_lut;
+    _cacheParams.tf_minMax = minMax;
+}
+
+void ParticleRenderer::_renderParticles()
+{
+    typedef struct {
+        glm::vec3  p;
+        float v;
+    } Vertex;
+    vector<Vertex> vertices;
+    vector<int>    sizes;
+    vector<Vertex> sv;
+
+    int nStreams = _particles.size();
+    for (int s = 0; s < nStreams; s+=2) {
+        glm::vec4 p = _particles[s];
+        glm::vec4 p2 = _particles[s+1];
+        const vector<flow::Particle> stream = {flow::Particle(p[0],p[1],p[2],0.,p[3]), flow::Particle(p2[0],p2[1],p2[2],0.,p2[3])};
```

Comment:
You already know that this vector has a size of 2, and the size won't change. Why not use an std::array?

Reply:
Makes sense.  Fixed.